File: internal/interpreters/r_requires.go
package interpreters

// Copyright (C) 2025 by Posit Software, PBC.

import (
	"encoding/json"
	"errors"
	"regexp"
	"strings"

	"github.com/posit-dev/publisher/internal/util"
)

type RProjectRRequires struct {
	ProjectPath util.AbsolutePath
}

func NewRProjectRRequires(projectPath util.AbsolutePath) *RProjectRRequires {
	return &RProjectRRequires{
		ProjectPath: projectPath,
	}
}

// Find the R version requested by the project if specified in any of the
// supported metadata files. The order of precedence is:
// 1. DESCRIPTION file (i.e. Depends: R (>= 3.5.0)
// 2. renv.lock
//
// The version specifications are the one defined by PEP 440
// if no version is found, an error is returned.
func (p *RProjectRRequires) GetRVersionRequirement() (string, error) {
	if version, err := p.readDescriptionFile(); err == nil && version != "" {
		return version, nil
	}
	if version, err := p.readRenvLockl(); err == nil && version != "" {
		return version, nil
	}
	return "", errors.New("no R version requirement found")
}

// Read a DESCRIPTION file and look for the "Depends:" section.
// if the section contains an entry for R (IE: "R (>= 3.5.0)")
// return the requirement constraint.
func (p *RProjectRRequires) readDescriptionFile() (string, error) {
	path := p.ProjectPath.Join("DESCRIPTION")
	content, err := path.ReadFile()
	if err != nil {
		return "", err
	}

	lines := strings.Split(string(content), "\n")

	// Find the "Depends:" line and extract all the dependencies.
	var deps []string
	found := false
	for _, line := range lines {
		if strings.HasPrefix(line, "Depends:") {
			deps = append(deps, strings.TrimPrefix(line, "Depends:"))
			found = true
		} else if found && (strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t")) {
			deps = append(deps, strings.TrimSpace(line))
		} else if found {
			break
		}
	}

	// Join the dependencies into a single string and look for R.
	// The regex will match "R (ANYTHING)"
	all := strings.Join(deps, " ")
	re := regexp.MustCompile(`\bR\s*\(([^)]+)\)`)
	match := re.FindStringSubmatch(all)
	if len(match) >= 2 {
		return match[1], nil
	}
	return "", nil
}

// Read a renv.lock file and return the version string.
// The file is a JSON file that contains a set of dependencies.
//
//	"R": {
//		 "Version": "4.2.2"
//	}, ...
func (p *RProjectRRequires) readRenvLockl() (string, error) {
	path := p.ProjectPath.Join("renv.lock")
	data, err := path.ReadFile()
	if err != nil {
		return "", err
	}

	var parsed map[string]any
	if err := json.Unmarshal(data, &parsed); err != nil {
		return "", err
	}

	if rSection, ok := parsed["R"].(map[string]any); ok {
		if version, ok := rSection["Version"].(string); ok {
			return "~=" + version, nil
		}
	}
	return "", nil
}

File: extensions/vscode/src/interpreters/rRequires.ts
// Copyright (C) 2026 by Posit Software, PBC.

import { readFile } from "node:fs/promises";
import path from "node:path";
import { adaptToCompatibleConstraint } from "./versionConstraints";

/**
 * Find the R version requested by the project, checking in order:
 * 1. DESCRIPTION file (Depends: R (>= x.y.z))
 * 2. renv.lock (R.Version)
 *
 * Returns the version specification, or empty string if not found.
 */
export async function getRRequires(projectDir: string): Promise<string> {
  const fromDescription = await readDescriptionFile(projectDir);
  if (fromDescription) {
    return fromDescription;
  }

  const fromRenvLock = await readRenvLock(projectDir);
  if (fromRenvLock) {
    return fromRenvLock;
  }

  return "";
}

async function readFileText(filePath: string): Promise<string | null> {
  try {
    return await readFile(filePath, "utf-8");
  } catch {
    return null;
  }
}

/**
 * Read DESCRIPTION file and look for R version in the Depends: section.
 * Matches patterns like "R (>= 3.5.0)".
 */
async function readDescriptionFile(projectDir: string): Promise<string | undefined> {
  const content = await readFileText(path.join(projectDir, "DESCRIPTION"));
  if (content === null) {
    return undefined;
  }

  const lines = content.split("\n");
  const deps: string[] = [];
  let found = false;

  for (const line of lines) {
    if (line.startsWith("Depends:")) {
      deps.push(line.substring("Depends:".length));
      found = true;
    } else if (found && (line.startsWith(" ") || line.startsWith("\t"))) {
      deps.push(line.trim());
    } else if (found) {
      break;
    }
  }

  const all = deps.join(" ");
  const re = /\bR\s*\(([^)]+)\)/;
  const match = re.exec(all);
  if (match && match[1]) {
    return match[1];
  }

  return undefined;
}

/**
 * Read renv.lock (JSON) and extract R.Version, converting it to a
 * compatible constraint via adaptToCompatibleConstraint.
 */
async function readRenvLock(projectDir: string): Promise<string | undefined> {
  const content = await readFileText(path.join(projectDir, "renv.lock"));
  if (content === null) {
    return undefined;
  }

  try {
    const parsed = JSON.parse(content);
    const version = parsed?.R?.Version;
    if (typeof version === "string" && version) {
      return adaptToCompatibleConstraint(version);
    }
  } catch {
    // Invalid JSON, ignore
  }

  return undefined;
}
